nalgebra-sparse/src/io/matrix_market.rs

```diff
@@ -1329,3 +1417,110 @@ fn next_dense_coordinate(
         }
     }
 }
+
+/// Write a sparse matrix into Matrix Market format string.
+///
+/// Our exporter only writes matrix into `coordiante` and `general` format.
+///
+///
+/// Examples
+/// --------
+/// ```
+/// # use matrixcompare::assert_matrix_eq;
+/// use nalgebra_sparse::io::{write_to_matrix_market_str,load_coo_from_matrix_market_str};
+/// let str = r#"
+/// %%matrixmarket matrix coordinate integer general
+/// 5 4 2
+/// 1 1 10
+/// 2 3 5
+/// "#;
+/// let matrix = load_coo_from_matrix_market_str::<i32>(&str).unwrap();
+/// let generated_matrixmarket_string = write_to_matrix_market_str(&matrix);
+/// // 'generated_matrixmarket' should equal to the 'matrix'
+/// let generated_matrixmarket = load_coo_from_matrix_market_str::<i32>(&generated_matrixmarket_string).unwrap();
+/// assert_matrix_eq!(matrix,generated_matrixmarket);
+/// ```
+pub fn write_to_matrix_market_str<T: MatrixMarketScalar, S: MatrixMarketExport<T>>(
+    sparse_matrix: &S,
+) -> String {
+    let mut bytes = Vec::<u8>::new();
+    // This will call impl<A: Allocator> Write for Vec<u8, A>
+    // The vector will grow as needed.
+    // So, unwrap here won't cause any issue.
+    write_to_matrix_market(&mut bytes, sparse_matrix).unwrap();
+    // safety issue is because 'from_utf8_unchecked' does not check that the bytes passed to it are valid UTF-8.
+    // Since 'bytes' created here is valid UTF-8 data, so it will not cause any problem.
+    unsafe { String::from_utf8_unchecked(bytes) }
```

Comment:
Because everything, which will be written, is controlled by the code, not the user. And from the code side, we only write header, one line of comment and data, which  is only implemented for basic data types, e.g. int, float. So that's why I think it should be safe to use `from_utf8_unchecked`, 

however, there is another 2 safe operations, one is [from_utf8](https://doc.rust-lang.org/stable/std/string/struct.String.html#method.from_utf8), which gives an error when non-utf8 data encountered, another one is [from_utf8_lossy](https://doc.rust-lang.org/stable/std/string/struct.String.html#method.from_utf8_lossy), which converts non-utf8 character to `std::char::REPLACEMENT_CHARACTER`, it looks like this �

so I don't know which one you think would be better

Reply:
I agree that with the current implementation, it's probably fine. Basically as long as we're using `write!` everywhere we should be in the clear. But we'd also like code to be robust to later changes, and this is a very implicit invariant that might not be obvious for someone working on the code base later on.

I think `from_utf8(bytes).expect("..")` is probably the way to go: after all, our code *should* be unicode-compatible, so if it's not then there must be a bug in our program.